docs/RFCS/raft_consistency_checker.md

```diff
@@ -0,0 +1,114 @@
+- Feature Name: Raft consistency checker
+- Status: draft
+- Start Date: 2016-02-10
+- Authors: Ben Darnell, Bram Gruneir, Vivek Menezes
+- RFC PR: (PR # after acceptance of initial draft)
+- Cockroach Issue: #837
+
+# Summary
+
+An online system that periodically collects a SHA on the range replicas to
+check for consistency; a snapshot for a range at a specific point in the raft log
+should have the same SHA on all replicas. An API to be used in tests to check for
+consistency problems will also be provided.
+
+# Motivation
+
+Consistency! Correctness at scale.
+
+# Detailed design
+
+The proposal is for each range replica leader to periodically collect a SHA of
+each replica to keep track of consistency problems and report them. The
+consistency checker runs in three phases:
+
+1. All replicas agree on the snapshot at which a SHA is to be computed.
+2. A SHA is computed on all replicas.
+3. The leader collects the SHA from all replicas and reports any replica
+computing a different SHA.
+
+The above is implemented by introducing two raft commands in
+roachpb.RequestUnion:
+
+1. ComputeChecksum: A replica on receiving this command will reply
+immediately, and asynchronously compute the SHA for the current snapshot. It
+will continue executing commands while it computes the SHA on the side. The
+command will contain a UUID and a version number identifying the method
+used in computing the SHA. The SHA is reported back to the leader through the
+next command.
+2. CollectChecksum: The leader will supply a UUID for the ComputeChecksum
```

Comment:
I think there is value in a replica knowing that there is a conflict.  If you have an extremely slow store, say due to bad sectors, unless the leader keeps polling and there are no leadership changes, it might be missed entirely.  Unlikely, but possible.  It's a small amount of info to send as part of the command and in the least allows us to throw a warning that something is wrong.